kedro/io/kedro_data_catalog.py

```diff
@@ -551,11 +551,100 @@ def add(
             )
         self.__setitem__(ds_name, dataset)
 
+    def filter(
+        self,
+        name_regex: str | None = None,
+        name_regex_flags: int | re.RegexFlag = re.IGNORECASE,
+        type_regex: str | None = None,
+        type_regex_flags: int | re.RegexFlag = 0,
+        by_type: type | list[type] | None = None,
+    ) -> List[str]:  # noqa: UP006
+        """Filter dataset names registered in the catalog based on name and/or type.
+
+        This method allows filtering datasets by their names and/or types using optional
+        regular expression patterns. Each pattern can also support optional regex flags
+        for customization. If no filters are provided, all dataset names are returned.
+
+        Args:
+            name_regex: Optional regular expression to filter dataset names by name.
+            name_regex_flags: Optional regex flags for the name filter.
+                By default, IGNORECASE key is set.
+            type_regex: Optional regular expression to filter dataset names by their type.
+                The provided regex is matched against the full dataset type path, for example:
+                `kedro_datasets.pandas.parquet_dataset.ParquetDataset`.
+            type_regex_flags: Optional regex flags for the type filter.
+            by_type: Optional dataset types to filter by.
```

Comment:
I think it would be helpful to give an example of a `by_type` argument. And maybe clarify that this is an instance of type of check instead of regex match? And also clarify it can take a single type or list. I know people can also read that from the function signature, but better to be extra clear IMO.